File: Array/1329.Sort_the_Matrix_Diagonally/Python/1329.Sort_the_Matrix_Diagonally.py
class Solution(object):
    def diagonalSort(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: List[List[int]]
        """
        #===============================#
        # Greedy-based traversal method #
        #===============================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_columns = len(mat), len(mat[0])

        ##### Record dictionary/hashmap #####
        record_dict = {}


        ##############################################################
        #Greedy-based loop traversal with recorded dictionary/hashmap
        ##### Step 1: Record indexed-value with dictionary/hashmap #####
        for rows_idx in range(len_rows):
            for columns_idx in range(len_columns):
                curr_diff = (rows_idx - columns_idx) #Current differences

                ##### Check if the current indexed-value existed or not #####
                if (curr_diff not in record_dict):
                    record_dict[curr_diff] = [(mat[rows_idx])[columns_idx]] #Keep updating/recording
                else:
                    (record_dict[curr_diff]).append((mat[rows_idx])[columns_idx]) #Keep updating/recording

        ##### Step 2: Sorted operations #####
        for sub_key in record_dict:
            (record_dict[sub_key]).sort(reverse = True) #Keep updating/overwriting

        ##### Step 3: Looped-traversal with recorded dictionary
        for rows_idx in range(len_rows):
            for columns_idx in range(len_columns):
                (mat[rows_idx])[columns_idx] = (record_dict[(rows_idx - columns_idx)]).pop((-1)) #Keep updating/overwriting

        return mat

File: Array/1329.Sort_the_Matrix_Diagonally/C++/1329.Sort_the_Matrix_Diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>> & mat) 
    {
        //===========================================//
        // Input type:                               //
        //   - vector<vector<int>> ref. (mat vector) //
        // Return type:                              //
        //   - vector<vector<int>> (result vector)   //
        //===========================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = mat.size(), len_columns = (mat[0]).size();

        ///// Record hashmap, dictionary /////
        unordered_map<int, vector<int>> record_map;

        ///// Record differences /////
        int record_diff;


        /*Greedy-based loop traversal with recorded hashmap, dictionary*/
        ///// Step 1: Record indexed-value with hashmap, dictionary /////
        for (int rows_idx = 0; (rows_idx < len_rows); (rows_idx++)) //Row
        {
            for (int columns_idx = 0; (columns_idx < len_columns); (columns_idx++)) //Column
            {
                (record_map[(rows_idx - columns_idx)]).push_back((mat[rows_idx])[columns_idx]); //Keep updating, recording

            } //Column

        } //Row

        ///// Step 2: Sorted operations /////
        for (auto sub_item = record_map.begin(); (sub_item != record_map.end()); (sub_item++)) //Whole
        {
            sort((sub_item->second).begin(), (sub_item->second).end(), std::greater<int>()); //Keep updating, overwriting 

        } //Whole

        ///// Step 3: Looped-traversal with recorded hashmap /////
        for (int rows_idx = 0; (rows_idx < len_rows); (rows_idx++)) //Row
        {
            for (int columns_idx = 0; (columns_idx < len_columns); (columns_idx++)) //Column
            {
                record_diff = (rows_idx - columns_idx); //Record differences

                (mat[rows_idx])[columns_idx] = (record_map[record_diff]).back(); //Keep updating, overwriting
                (record_map[record_diff]).pop_back(); //Keep updating, popped

            } //Column

        } //Row

        return mat;
    }
};
